### PR TITLE
feat: Hugging Face Pipeline as a new Chat Model

### PR DESCRIPTION
**Pull request by eryk-dsai**
## Description:

This Pull Request adds **Hugging Face Pipeline** as a new LangChain Chat Model. 

It effectively allows the user to use any instruction tuned model with LangChain's Chat interface with a newly added Chat Templates [link](https://huggingface.co/blog/chat-templates) [link](https://huggingface.co/docs/transformers/main/chat_templating)

For example, it is now possible to define the HF Pipeline with the `meta-llama/Llama-2-7b-chat-hf` model and [arbitrary QLoRA configuration](https://huggingface.co/blog/4bit-transformers-bitsandbytes) (as shown in the [example](https://github.com/deepsense-ai/langchain/blob/llama2-chat/docs/extras/integrations/chat/huggingface_llama2.ipynb)) and run it locally with LangChain on a GPU with less than 6GB of VRAM. 

@deepsense_ai

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
naming should reflect that it uses huggingface (there are plenty of other llama implementations that use something else

**Comment by rlancemartin** [comment]
Nice addition! @jacoblee93 recently added chat model for Ollama, so may be able to have a quick look at this.

File: libs/langchain/langchain/chat_models/huggingface_llama2.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        return values
+
+    @staticmethod
+    def format_messages_as_text(messages: List[BaseMessage]) -> str:
+        """
+        Transform List of Chat Messages to text following Meta's prompt guidelines.
+
+        Prompt template with System Message:
+        ```
+        <s>[INST] <<SYS>>
```
Comment (jacoblee93) [inline libs/langchain/langchain/chat_models/huggingface_llama2.py:61]: Interesting, this probably means we need to update:

https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/chat_models/ollama.py

File: libs/langchain/langchain/chat_models/huggingface_llama2.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+class SystemTokens(Enum):
+    def __str__(self) -> str:
+        return self.value
+
+    B_SYS = "<<SYS>>"
+    E_SYS = "<</SYS>>"
+
+
+class ChatLlama2Hf(BaseChatModel):
```
Comment (jacoblee93) [inline libs/langchain/langchain/chat_models/huggingface_llama2.py:36]: What about making this a generic `ChatHuggingFacePipeline` class that could be expanded as more chat models become available? Will defer to others for scope

File: libs/langchain/langchain/chat_models/huggingface_llama2.py (hunk 1 above)
Comment (jacoblee93) [reply libs/langchain/langchain/chat_models/huggingface_llama2.py:61]: Managed to track down a source but might be nice to include this in the docs/in a comment here?

https://github.com/facebookresearch/llama-recipes/blob/df77625e48c3994aef19702fb331215f7fb83494/docs/inference.md?plain=1#L124

File: docs/extras/integrations/chat/huggingface_llama2.ipynb [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+   "metadata": {},
+   "source": [
+    "## Creating Hugging Face Pipeline instance:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following section loads the 7b version of the Llama 2 Chat model and uses the `bitandbytes` library to load a model in 4bit using NF4 quantization with double quantization and compute dtype bfloat16, which speeds up the underlying matrix multiplications.\n",
```
Comment (jacoblee93) [inline docs/extras/integrations/chat/huggingface_llama2.ipynb:108]: Typo: `bitsandbytes`, not `bitandbytes`

File: docs/extras/integrations/chat/huggingface_llama2.ipynb [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    "    torch_dtype=torch.float16,\n",
+    "    device_map=\"auto\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Defining LangChain Llama-2-Chat model:"
```
Comment (jacoblee93) [inline docs/extras/integrations/chat/huggingface_llama2.ipynb:199]: Missing `a`, should be: `Defining a LangChain Llama-2-Chat model`

**Comment by jacoblee93** [comment]
Left some comments, looks fine to me generally but will defer to more experienced maintainers for final review.

File: libs/langchain/langchain/chat_models/huggingface_llama2.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        if importlib.util.find_spec("transformers") is not None:
+            from transformers import StoppingCriteria, StoppingCriteriaList
+
+        device = self.pipeline.device.type
+        if device == "cuda":
+            # in the multi-gpu case, stopping criteria tokens
+            # need to be on the same device:
+            device = f"{device}:{self.pipeline.device.index}"
+
+        class StoppingCriteriaSub(StoppingCriteria):
```
Comment (jacoblee93) [inline libs/langchain/langchain/chat_models/huggingface_llama2.py:126]: Naming could be better here

File: libs/langchain/langchain/chat_models/huggingface_llama2.py (hunk 1 above)
Comment (eryk-dsai) [reply libs/langchain/langchain/chat_models/huggingface_llama2.py:61]: Yes, I think so. I'll include in the docstring

**Comment by eryk-dsai** [comment]
Hi @jacoblee93 

I implemented your suggestions

**Comment by jokokojote** [comment]
Hey, I stumbled upon this, because I had the need to have a local transforms chat pipeline myself. I wrote something similar (but more basic), because I did not find anything in the package. Of course I should have checked ongoing pull requests first before start implementing something myself. ^^

However, I like what you did @eryk-dsai , but wonder if it would be better to leverage the given `HuggigFacePipeline` implementation from `langchain_community`, to avoid double and deviant code. That is, the chat implementation would be more kind of a wrapper. That would seem logical to me, because in the end it does not provide anything else than incorporation of the huggingface chat templates. It is done this way in `langchain_community`multiple times e.g. for `ChatHuggingFace` and `HuggingFaceTextGenInference`, `HuggingFaceEndpoint`, `HuggingFaceHub`. 

E.g. like this (basic example of the concept):
```
class ChatHuggingFacePipeline(SimpleChatModel ):

    llm: HuggingFacePipeline # no direct transformers pipeline, but given HuggingFacePipeline implementation

    def __init__(self, **kwargs: Any):
        super().__init__(**kwargs)

    def _call(
        self,
        messages: List[BaseMessage],
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> str:
        
        messages = langchain_to_transformers_msgs(messages)
        input = self.llm.pipeline.tokenizer.apply_chat_template(messages, tokenize=False)

        out = self.llm._generate([input], stop=stop, run_manager=run_manager, **kwargs)

        return out.generations[0][0].text
    
    @property
    def _llm_type(self) -> str:
        """Return type of chat model."""
        return "huggingface_pipeline_chat_wrapper"
```

So, leaving the pipeline instantiation etc. on transformers level to the `HuggingFacePipeline`class.

One more suggestion from my point one view: One reason I found this PR (to) late is, that I checked for such an implementation in `langchain_community`, because `HuggingFacePipeline` is there, as well as all the other huggingface stuff (`ChatHuggingFace` etc.) So I am wondering if this should be moved to `langchain_community` as well, for consistency.

**Comment by raimannma** [comment]
Worked perfectly for me, I can now use Llama 3 with Agents. Without this I couldn't get it to work.

Thank you very much @eryk-dsai

**Comment by hwchase17** [comment]
huggingface now its own package

**Comment by djstrong** [comment]
What do you mean? It is still not implemented.

**Pull request closed without merging**